### ISSUE TITLE
Need to respect image registry mirror configuration for direct image pulls

### ISSUE THREAD

**Issue by brandond** [rancher/rke2#233]
tl;dr We should respect the settings from registries.yaml when doing direct image pulls.

There are a few places in the bootstrap code path where rke2 directly pulls images if they cannot be found locally. Example: https://github.com/rancher/rke2/blob/master/pkg/bootstrap/stage.go#L58

These image pulls do not currently respect private registry configuration (registries.yaml) since this file is only used to configure containerd. This means that if the bootstrap content (images, root, charts, etc) cannot be found locally, rke2 goes directly out to docker.io. If it cannot reach docker.io, it fails.

We're currently pulling images via go-containerregistry, we might need to hack something in to the [transport](https://github.com/google/go-containerregistry/tree/master/pkg/v1/remote/transport) if it doesn't have any direct mirror/endpoint override support.

**Comment by nikkelma** [comment]
TLDR: Based on my research, `go-containerregistry` directly links image namespace (`docker.io`, `quay.io`, `my-internal-registry.com`) and registry hostname, preventing easy registry mirroring. Enabling/allowing mirroring would either require adding options and functionality to `go-containerregistry` or would require switching container registry clients. From what I can tell, the transport is used only for low-level HTTP authentication and doesn't support redirecting requests, unsure if this would provide a solution.

I'll be following this closely as I'm using `go-containerregistry` for helping with air-gap installs and wanting the option for this behavior. I took a look at how other projects handled this, specifically moby [1], containerd [2], and cri-o [3]. Moby hasn't yet added registry mirror support, but containerd and cri-o have.

The best summary of `go-containerregistry` is introduced in the Moby PR [here](https://github.com/moby/moby/pull/34319#issuecomment-647200863):
> ... the original sin of conflating image namespace with registry hostname ...

`go-containerregistry` currently has a one-to-one mapping between image namespace and registry hostname. For this to change, some sense of registry mirrors would need to be implemented (unless I'm missing something).

One path is to design and implement of registry mirroring inside of `go-containerregistry`, whether in a fork or directly upstream. Seeing that the moby PR has been open for 3 years with SUSE backing, don't think this can be assumed to be easy.

The other path is to switch to a beefier registry client that does support mirroring, and importing the `registries.yaml` into this client's mirror config.

I don't think `go-containerregistry`'s `remote.WithTransport()` is a possible solution, as `go-containerregistry/pkg/v1/remote/transport/` is described as:
> This package implements an http.RoundTripper that transparently performs:
> - Token Authentication and
> - OAuth2 Authentication
so as I understand it, this should be used mainly for separating authentication from registry config.


[1] https://github.com/moby/moby/pull/34319
[2] https://github.com/containerd/cri/pull/531
[3] https://github.com/containers/image/pull/564

**Comment by cjellick** [comment]
This issue is superseded by https://github.com/rancher/rke2/issues/271. The crux of it is “do something fancy to get our go-client to respect what is in registries.yaml.”

But 271 is the design for `default-system-registry`.

I think the impl of 271 and the alternative to use the tarball images is enough of a solution  such that this issue is wholly unnecessary.
